devito/interfaces.py

```diff
@@ -72,7 +73,20 @@ def __new__(cls, *args, **kwargs):
     def __init__(self):
         """Initialise from a cached instance by shallow copying __dict__."""
         original = _SymbolCache[self.__class__]
-        self.__dict__ = original.__dict__.copy()
+        self.__dict__ = original().__dict__.copy()
+
+    def clean(self):
+        """Cleans up a :class:`SymbolicData` object by removing
+        it from the symbolic cache.
+        """
+        if self._cached():
+            self._clean()
```

Comment:
Ok, so I don't think that's true. I just build a custom `__del__` on `_SymbolicData` locally and it works fine with your test, provided I clear the sympy cache properly.